apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/ClientPool.cs

```diff
@@ -85,53 +108,182 @@ public override int GetHashCode()
+                !_creationTask.Value.IsFaulted &&
+                ReferenceEquals(_creationTask.Value.ResultWithUnwrappedExceptions(), client);
+
+            private async Task<ChannelCredentials> CreateDefaultChannelCredentialsAsync()
+            {
+                var appDefaultCredentials = await GoogleCredential.GetApplicationDefaultAsync().ConfigureAwait(false);
+                if (appDefaultCredentials.IsCreateScopedRequired)
+                {
+                    appDefaultCredentials = appDefaultCredentials.CreateScoped(SpannerClient.DefaultScopes);
+                }
+                return appDefaultCredentials.ToChannelCredentials();
+            }
+
+            private async Task<SpannerClient> AcquireClientImplAsync()
+            {
+                var endpoint = _parentKey.Endpoint;
+                ChannelCredentials channelCredentials;
+
+                if (_parentKey.Credential == null)
+                {
+                    channelCredentials = await CreateDefaultChannelCredentialsAsync().ConfigureAwait(false);
+                }
+                else
+                {
+                    channelCredentials = _parentKey.Credential.ToChannelCredentials();
+                }
+
+                var channel = new Channel(
+                    endpoint.Host,
+                    endpoint.Port,
+                    channelCredentials);
+                Logger.LogPerformanceCounterFn("SpannerClient.RawCreateCount", (x) => x + 1);
+
+                return SpannerClient.Create(channel);
+            }
+
+            public async Task<SpannerClient> AcquireClientAsync()
+            {
+                if (_creationTask.Value.IsFaulted)
+                {
+                    //retry an already failed task.
+                    _creationTask = new Lazy<Task<SpannerClient>>(AcquireClientImplAsync);
+                }
+
+                var spannerClient = await _creationTask.Value.ConfigureAwait(false);
+                Interlocked.Increment(ref _refCount);
+                OnPriorityChanged();
+
+                return spannerClient;
+            }
+
+            public void Release()
+            {
+                Interlocked.Decrement(ref _refCount);
+                OnPriorityChanged();
+            }
+
+            /// <inheritdoc />
+            public int CompareTo(SpannerClientCreator other)
+            {
+                if (ReferenceEquals(this, other))
+                {
+                    return 0;
+                }
+                if (ReferenceEquals(null, other))
+                {
+                    return 1;
+                }
+                int refCountComparison = RefCount.CompareTo(other.RefCount);
+                if (refCountComparison != 0)
+                {
+                    return refCountComparison;
+                }
+                // This has a chance of returning 0 even if the object is different.
+                // That is fine and is handled by the PriorityHeap.
```

Comment:
Is it actually handled by PriorityList? Won't it treat the two values as equal, and not allow the second to be added?